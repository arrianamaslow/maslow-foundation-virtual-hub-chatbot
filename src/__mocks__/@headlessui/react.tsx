/**
 * @jest-environment jsdom
 */

import { DialogProps } from '@headlessui/react'
import React, { ReactNode } from 'react'

type DialogPanelTestProps = {
  children?: ReactNode
}

interface DialogTestProps extends DialogProps {
  children?: ReactNode
}

const DialogPanel: React.FC<DialogPanelTestProps> = ({ children }) => {
  return <div data-testid='mock-dialog-panel-component'>{children}</div>
}

const Dialog: React.FC<DialogTestProps> = ({ open, children }) => {
  return (
    <div data-testid='mock-dialog-component'>
      {open ? 'DIALOGUEOPENED' : 'DIALOGUECLOSED'}
      {children}
    </div>
  )
}

export { Dialog, DialogPanel }
