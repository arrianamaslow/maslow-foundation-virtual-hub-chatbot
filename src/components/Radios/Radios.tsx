import React, { useEffect, useState } from 'react'
import CheckBox from '@/components/CheckBox/CheckBox'

export type RadiosProps = {
  radioGroupName: string
  radioValues: string[]
  defaultSelectedIndex?: number
  radioClickedCallback: React.Dispatch<React.SetStateAction<string>>
}

const Radios: React.FC<RadiosProps> = ({
  radioGroupName,
  radioValues,
  defaultSelectedIndex,
  radioClickedCallback
}) => {
  const [selectedIndex, setSelectedIndex] = useState<number | undefined>(defaultSelectedIndex)

  const onRadioChange = (newSelectedIndex: number) => {
    setSelectedIndex(newSelectedIndex)
    radioClickedCallback(radioValues[newSelectedIndex])
  }

  useEffect(() => {
    if (
      !(defaultSelectedIndex === undefined) &&
      defaultSelectedIndex >= 0 &&
      defaultSelectedIndex < radioValues.length
    ) {
      onRadioChange(defaultSelectedIndex)
    }
  }, [])

  return (
    <div className='flex md:h-[20px] min-w-[8rem]'>
      {radioValues.map((radioValue, index) => (
        <div key={index} className=' min-w-[8rem]'>
          <label
            tabIndex={index}
            className='flex gap-2 items-center md:h-[20px]'
            data-testid={`mock-radio-option-${index}`}
          >
            <CheckBox tickIsActive={selectedIndex === index} />
            <input
              type='radio'
              name={radioGroupName}
              className='hidden'
              onChange={(e) => {
                onRadioChange(index)
              }}
              checked={selectedIndex === index}
            />
            {radioValue}
          </label>
        </div>
      ))}
    </div>
  )
}

export default Radios
