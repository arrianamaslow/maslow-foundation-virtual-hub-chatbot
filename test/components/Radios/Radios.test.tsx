/**
 * @jest-environment jsdom
 */

import { fireEvent, render, screen } from '@testing-library/react'
import Radios, { RadiosProps } from '@/components/Radios/Radios'
import '@testing-library/jest-dom'

jest.mock('@/components/CheckBox/CheckBox', () => {
  const MockCheckBox = ({ tickIsActive }: { tickIsActive: boolean }) => {
    return <p> {tickIsActive ? 'CHECKENABLED' : 'CHECKDISABLED'}</p>
  }
  MockCheckBox.displayName = 'MockCheckBox'
  return MockCheckBox
})

const defaultProps: RadiosProps = {
  radioValues: ['hello', 'world', 'everyone'],
  radioGroupName: 'test',
  radioClickedCallback: () => {}
}

describe('<Radios />', () => {
  it('renders correctly', () => {
    const { container } = render(<Radios {...defaultProps} />)

    expect(container).toMatchSnapshot()
  })

  it('renders one selected radio if given a default to tick', () => {
    render(<Radios {...defaultProps} defaultSelectedIndex={1} />)

    const radioButtonOne = screen.getByTestId('mock-radio-option-0')
    const radioButtonTwo = screen.getByTestId('mock-radio-option-1')
    const radioButtonThree = screen.getByTestId('mock-radio-option-2')

    expect(radioButtonOne).toHaveTextContent('CHECKDISABLED')
    expect(radioButtonTwo).toHaveTextContent('CHECKENABLED')
    expect(radioButtonThree).toHaveTextContent('CHECKDISABLED')
  })

  it('renders a un-selected radio when defaultSelectedIndex is not set', () => {
    render(<Radios {...defaultProps} />)

    const checkEnabled = screen.queryByText('CHECKENABLED')
    expect(checkEnabled).toBeFalsy()
  })

  it('changes selected radio on button on click', () => {
    render(<Radios {...defaultProps} defaultSelectedIndex={1} />)

    const radioButtonOne = screen.getByTestId('mock-radio-option-0')
    const radioButtonTwo = screen.getByTestId('mock-radio-option-1')
    const radioButtonThree = screen.getByTestId('mock-radio-option-2')

    expect(radioButtonOne).toHaveTextContent('CHECKDISABLED')
    expect(radioButtonTwo).toHaveTextContent('CHECKENABLED')
    expect(radioButtonThree).toHaveTextContent('CHECKDISABLED')

    fireEvent.click(radioButtonOne)

    expect(radioButtonOne).toHaveTextContent('CHECKENABLED')
    expect(radioButtonTwo).toHaveTextContent('CHECKDISABLED')
    expect(radioButtonThree).toHaveTextContent('CHECKDISABLED')
  })

  it('calls callback function on button click', () => {
    const callbackCheck = jest.fn()
    render(<Radios {...defaultProps} radioClickedCallback={callbackCheck} />)

    const firstButton = screen.getByTestId('mock-radio-option-0')
    fireEvent.click(firstButton)

    expect(callbackCheck).toHaveBeenCalledWith('hello')

    const secondButton = screen.getByTestId('mock-radio-option-1')
    fireEvent.click(secondButton)

    expect(callbackCheck).toHaveBeenCalledWith('world')
  })

  it('callback function called on initialisation', () => {
    const callbackCheck = jest.fn()
    render(
      <Radios {...defaultProps} defaultSelectedIndex={1} radioClickedCallback={callbackCheck} />
    )

    expect(callbackCheck).toHaveBeenCalledWith('world')
  })

  it('renders four radio buttons when four radioValues are passed', () => {
    render(
      <Radios
        {...defaultProps}
        radioValues={['go falafel', 'rustica', 'bundobust', 'northern monk']}
        defaultSelectedIndex={1}
      />
    )

    const radioButtons = screen.getAllByRole('radio')
    expect(radioButtons.length).toBe(4)
  })
})
