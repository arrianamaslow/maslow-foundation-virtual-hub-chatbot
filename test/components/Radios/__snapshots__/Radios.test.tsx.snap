// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radios /> renders correctly 1`] = `
<div>
  <div
    class="flex md:h-[20px] min-w-[8rem]"
  >
    <div
      class=" min-w-[8rem]"
    >
      <label
        class="flex gap-2 items-center md:h-[20px]"
        data-testid="mock-radio-option-0"
        tabindex="0"
      >
        <p>
           
          CHECKDISABLED
        </p>
        <input
          class="hidden"
          name="test"
          type="radio"
        />
        hello
      </label>
    </div>
    <div
      class=" min-w-[8rem]"
    >
      <label
        class="flex gap-2 items-center md:h-[20px]"
        data-testid="mock-radio-option-1"
        tabindex="1"
      >
        <p>
           
          CHECKDISABLED
        </p>
        <input
          class="hidden"
          name="test"
          type="radio"
        />
        world
      </label>
    </div>
    <div
      class=" min-w-[8rem]"
    >
      <label
        class="flex gap-2 items-center md:h-[20px]"
        data-testid="mock-radio-option-2"
        tabindex="2"
      >
        <p>
           
          CHECKDISABLED
        </p>
        <input
          class="hidden"
          name="test"
          type="radio"
        />
        everyone
      </label>
    </div>
  </div>
</div>
`;
